src/account/PluginManagerInternals.sol

```diff
@@ -886,7 +888,20 @@ abstract contract PluginManagerInternals is IPluginManager {
 
         delete _storage.pluginData[oldPlugin];
 
-        // @TODO: Apply Data migration
+        // Retrieve data from the old plugin for migration
+        bytes memory migrationData = IPlugin(oldPlugin).getDataForMigration();
+
+        // Pass the migration data to the new plugin
+        try IPlugin(newPlugin).onReplaceForNewPlugin(migrationData) {} 
+        catch (bytes memory revertReason){
+            revert onReplaceForNewPluginFailed(newPlugin, revertReason);
+        }
+
+        // Call the old plugin's clean-up function
+        try IPlugin(oldPlugin).onReplaceForOldPlugin() {} 
+        catch (bytes memory revertReason){
+            revert onReplaceForOldPluginFailed(oldPlugin, revertReason);
+        }
```

Comment:
I'm thinking of a case where `onReplaceForOldPlugin` works so the old plugin's data is cleaned up, but `onReplaceForNewPlugin` fails. I don't think there's any way to recover the plugin data in this case, what do you think?

Reply:
Also at first, I was thinking of the case in `onReplaceForOldPlugin` works so the data for migration is cleaned-up but `onReplaceForNewPlugin` fails. I did't think we could recover the plugin data for this case. 
However, this is my mistake because I just forgot 'all or nothing' in contracts at that time. In the case I mentioned, It will be reverted to the original state, right?

Reply:
> Ability to uninstall / replace a plugin is really important in the situation that a user installed a malicious plugin, which imo the onUninstall is designed not to revert (I think it's similar to emergency exit or something). The onInstall and onReplaceForNewPlugin reverts when it fails, since it's doing an intial setup necessary for plugins.

I got what you mean and changed the sequence.

Reply:
I mean the logic of replacing each data stored in the plugin e.g. userOpValidation function.